autoload/vimteractive.vim

```diff
@@ -14,53 +62,100 @@ endfunction
         echoerr "Your version of vim is not compiled with +terminal. Cannot use vimteractive"
         return
     endif
 
+	let l:bufnr = bufnr('%')
+	let l:terminal_type = a:terminal_type
 	if a:terminal_type ==# '-auto-'
 		let term_type = get(g:vimteractive_default_shells, &filetype, &filetype)
 		if has_key(g:vimteractive_commands, term_type)
 			let l:terminal = get(g:vimteractive_commands, term_type)
-			let g:current_term_type = term_type
+			let l:terminal_type = term_type
 		else
 			echoerr "Cannot determine terminal commmand for filetype " . &filetype
 			return
 		endif
 	else
 		let l:terminal = get(g:vimteractive_commands, a:terminal_type)
-		let g:current_term_type = a:terminal_type
 	endif
 
-    if g:vimteractive_terminal != '' && g:vimteractive_terminal != l:terminal
-        echoerr "Cannot run: " . l:terminal " Alreading running: " . g:vimteractive_terminal
-        return
+	" If we have already running Vimteractive terminal
+    if has_key(g:vimteractive_terminal, l:bufnr) && s:check_alive(l:bufnr)
+		if g:vimteractive_current_term_type[l:bufnr] == l:terminal
+			if bufwinnr(g:vimteractive_terminal[l:bufnr]) == -1
+				call s:show_term(l:bufnr)
+			else
+				echom "Terminal of type " . l:terminal_type . " already running for this buffer"
+			endif
+		else
+			echoerr "Cannot run: " . l:terminal
+				\ ". Already running: " . g:vimteractive_terminal[l:bufnr]
+		endif
+		return
     endif
 
-    if bufnr(g:vimteractive_buffer_name) == -1
-        " If no vimteractive buffer exists:
-        " Start the terminal
-        let job = term_start(l:terminal, {"term_name":g:vimteractive_buffer_name})
-        set nobuflisted                          " Unlist the buffer
-        set norelativenumber                     " Turn off line numbering if off
-        set nonumber                             " Turn off line numbering if off
-        wincmd p                                 " Return to the previous window
-        let g:vimteractive_terminal = l:terminal " Name the current terminal
-
-    elseif bufwinnr(g:vimteractive_buffer_name) == -1
-        " Else if vimteractive buffer not open:
-        " Split the window
-        split
-        " switch the top window to the vimteractive buffer
-        execute ":b " . g:vimteractive_buffer_name
-        " Return to the previous window
-        wincmd p
-
-    else
-        " Else if throw error
-        echoerr "vimteractive already open. Quit before opening a new buffer"
-    endif
+	let l:term_buffer_name = "term_" . l:terminal_type
+	let i = 1
+	while bufnr(l:term_buffer_name) != -1
+		let l:term_buffer_name = "term_" . l:terminal_type . '_' . i
+		let i += 1
+	endwhile
+
+	echom "Starting " . l:terminal
+	" Else we want to create a new term
+	call term_start(l:terminal, {
+		\ "term_name": l:term_buffer_name,
+		\ "term_finish": "close",
+		\ "term_kill": "term"
+		\ })
+
+	" Unlist the buffer
+	" set nobuflisted
+	" Turn line numbering off
+	set nonumber norelativenumber
```

Comment:
I think this should cover all cases, although I've had issues in the past with regards to relative numbering and vimteractive. Have you experienced anything similar?